rfc/rfc-79/rfc-79.md

```diff
@@ -0,0 +1,138 @@
+
+
+Here the major failure scenarios are round steps 5, 6, 7 and 8. If a spark task(or a subset of tasks) is attempted twice in step5, we could see 2 data files written to disk. And we should be able to find markers for both attempts. During marker based reconciliation, we should be able to account for the spurious data file and delete it.
+
+#### Case1:
+Stray executor:
+
+1 -> 2 … 5(writes) ->
+The write stage completes and spark triggers a kill for any additional tasks that is currently running. Lets say due to executor failure/spark speculative execution, another task was triggered, but attempt1 of task completed and is sent back to the driver when the write stage is completed. And hence we have a stray task running. At the end of the stage, spark will be triggering a kill for any stray tasks. And due to networking glitch, it could so happen that the task could not be reached for killing and it is running. During marker based reconciliation, we might see the marker file, but the data file could be missing if the task is still making progress. And eventually when the task completes, the spurious data file could be added to storage. Even the commit metadata may not be tracking this spurious data file.
+We heard from uber, that this stray executor could sometimes be running even after spark application is completed (very one off case)
+
+#### Case2:
+Dag retrigger while writing to MDT(spark rdd cache invalidation). Issue exacerbated with RLI.
+
+1 -> 2 … 5(writes) -> 6(marker based reconciliation) ->
+After marker based reconciliation is complete, while trying to write to MDT, say the dag is retriggered due to spark rdd cache invalidation. So, write handles could be re-executed, but the reconciliation may not get executed again. So, this could lead to duplicate data files on storage. Infact our commit metadata could be tracking one set of files, but MDT writes could be tracking a diff set of files.
+
+
+## Design
+
+We are proposing two additional markers to assist with detecting and thwarting spurious data files in Hudi on any cost.
+### 1. Completion Marker:
+   The marker mechanism discussed above strictly speaking is a begin marker. So, lets add a completion maker at the end of writes (write handle) for each data file produced to mark the completion of the write for a given task in addition to the begin marker. Completion marker will only track the fileId and commit time(write token should be ignored) and the writeStatus gets serialized as content. This might be in contrast to the begin markers where the marker will track the actual data file name.
+
+Write handle changes:
+In the  constructor
+1. If completion marker is already present for the given fileID and commit time, read the CM(completion marker), to deser write status and return right away. (short circuit).
+2. If completion marker is not present, go ahead w/ writing the data file(s).
+3. In the end just before returning the writeStatus back to driver:
```

Comment:
It's possible that this task may not delete the (data) file it created due to either
a) doing the "does completion marker exist" check right before the other task created the file
b) Encountering a transient failure before issuing the file delete or otherwise failing to delete the file
just to clarify, are we planning on handling/calling out this edge case in this RFC itself or is that something that will be handled in the implementation itself or future work (since I guess it would be unlikely edge case)?